src/submodule.c

```diff
@@ -628,8 +628,13 @@ int git_submodule_save(git_submodule *submodule)
 		(error = git_config_file_set_string(mods, key.ptr, submodule->url)) < 0)
 		goto cleanup;
 
-	if ((error = submodule_config_key_trunc_puts(&key, "branch")) < 0 ||
-		(error = git_config_file_set_string(mods, key.ptr, submodule->branch)) < 0)
+	if ((error = submodule_config_key_trunc_puts(&key, "branch")) < 0)
+		goto cleanup;
+	if (submodule->branch == NULL)
+		error = (error = git_config_file_delete(mods, key.ptr)) == GIT_ENOTFOUND ? 0 : error;
+	else
```

Comment:
This is too much gymnastics with the error code. I'd rather see the style we have in the rest of the libarary

```C
error = git_config_file_delete(...);
if (error == GIT_ENOTFOUND)
    error = 0;
```